### ISSUE TITLE
New training on Pong wont progress

### ISSUE THREAD

**Issue by ghost** [tambetm/simple_dqn#4]
Hi
After following the tutorial on the Nervana blog I started training a new Pong model. However even after 24 epochs the average reward was around random. I compared this to the checked in log file and saw that it started showing progress after only a few epochs. 

Here is the plot of relevant metrics.

![pong](https://cloud.githubusercontent.com/assets/6693858/12067258/b8db002e-afab-11e5-8965-6ea90c2e34df.png)

**Comment by tambetm** [comment]
Yes, this is a known issue with the latest Neon and it is taking longer to figure out than I expected. Please use the Neon commit https://github.com/NervanaSystems/neon/commit/7a56fa9645a51e97c05f2e5afbbd1df7057ae832 for training. That means in your Neon folder issue commands:

```
git checkout 7a56fa9645a51e97c05f2e5afbbd1df7057ae832 .
make
```

You also need to make a slight change in `src/deepqnetwork.py`, on line 30: change the `datatype =` to `default_dtype =`. Then it will work. But for filter visualization you still need the latest Neon.

I will put relevant notice in README too. The issue is tracked in https://github.com/NervanaSystems/neon/issues/153.

**Comment by ghost** [comment]
Thanks for the prompt reply!!

I am really impressed by your commitment to sharing knowledge and enabling others. I will try the fix and report back.

**Comment by tambetm** [comment]
You're welcome! Please let me know, if you have any interesting results. For example I've noticed, that using Adam optimizer (`--optimizer adam`) might give even better results and sooner. The default is `--optimizer rmsprop`.

**Comment by ghost** [comment]
Most definitely. Once I get the vanilla version working I am going to try some experiments. 

So I was reading about the convergence property of Q-Learning. I am very new to this stuff but is following intuition correct(I think you touched upon it in your blog post also):

"Q-Learning has been shown to be really useful. However it was not possible to apply the algorithm to problems with many states like video games. Combining Q-Learning with Neural Networks provides a scalable and robust representation of the Q table. This allows for tackling wide range of problems."

**Comment by ghost** [comment]
Its working now. I have started a parallel run with 'adam'. Will see how the metrics differ.

Thanks for your help. Have a great 2016

**Comment by tambetm** [comment]
@abhishekpatnia Yes, you got your intuition right. Because the neural network inputs can be any real numbers, the method can be applied even to the infinite state spaces. But notice, that the same doesn't apply to the actions - currently action space must be finite. See http://arxiv.org/abs/1509.02971 for extensions for continuous state spaces.

**Comment by guiambros** [comment]
@tambetm: just checking if this information is still up-to-date. At this moment I can't run fa9645a51 commit, as it doesn't have support for pre-Maxwell GPUs ("RuntimeError: Device 0 does not have CUDA compute capability 5.0 or greater").

**Comment by tambetm** [comment]
@guiambros I haven't had time to check the latest Neon, but about a month ago the problem was still there. You are welcome to test with latest Neon, the problem should be evident after about ~10 epochs, see https://github.com/NervanaSystems/neon/issues/153 for example plots.

**Comment by guiambros** [comment]
@tambetm: thanks! Just confirmed that the latest Neon (commit [66846b](https://github.com/NervanaSystems/neon/commit/66846b4097d256e176cd76559dfa4e0bc54ab6dc), Feb 9th) still has the same problem, just like you indicated on [neon#153](NervanaSystems/neon#153):
![breakout](https://cloud.githubusercontent.com/assets/205000/13041799/6f00f318-d388-11e5-81f2-0a60b04517e1.png)

I tried again with the commit [7a56fa](https://github.com/NervanaSystems/neon/commit/7a56fa9645a51e97c05f2e5afbbd1df7057ae832) and this time it worked well:

![breakout](https://cloud.githubusercontent.com/assets/205000/13041900/89709bb2-d389-11e5-8ea4-3185539f8e73.png)

**Comment by tambetm** [comment]
Thanks @guiambros for verifying this. Any hypotheses why that might happen? To me it seems, that target network fails first and starts returning zeros. Then loss for the main network goes up for a moment, but then the network learns to predict zero as well and loss returns to zero. For a long time I suspected saving-loading weights might cause the issue, but then I tried without target network (just set --target_steps 0) and it behaved exactly the same. Now I'm out of ideas...
